docs/Outputs.md

```diff
@@ -0,0 +1,184 @@
+# Flux Outputs
+
+This documentation contains the design of how Flux will produce outputs to external services.
+This doc has the aim of defining the tiny bits of Go functions that must be implemented to allow a pure Flux implementation for external service communication.
+ 
+## Endpoints, Encoders, and Transports
+
+There are three key components for an external service:
+ - an _endpoint_: it is the component that takes a table and pushes it to the external service.
+ - an _encoder_: it is the component that translates tables into something understandable by the external service.
+ - the _transport_: it is the communication protocol used to communicate with the external service.
+
+In a nutshell, endpoints use encoders to encode data and a transport to push data to external services.
+
+Encoders and transports are written in Go as `values.Function`s.
+For example there will be `json.encode`, as well as `csv.encode` for encoders, and `http.post` and `tcp.send` for transports.
+
+The minimal set that must be implemented is `http.post` and `json.encode`.
+
+Endpoints are implemented in pure Flux.
+
+## Flux
+
+Encoders are provided in Flux as `encode` functions.
+For example the JSON encoder is part of the `json` package.
+Encoders take an object `data` and return its encoding as a string:
+
+```flux
+encode: (data:object) -> string
+```
+
+Communication protocols are also implemented (in Go), so that one can:
+
+```flux
+import "http"
+import "tcp"
+
+import "json"
+import "csv"
+
+data = {a: 1, b: 2}
+
+json_enc = json.encode(data)
+csv_enc = csv.encode(data)
+
+http.post(header: {"Content Type": "application-json"}, url: "http://some/url/", data: json_enc)
+
+conn = tcp.dial(url: "tcp://some/url/")
+tcp.send(connection: conn, data: csv_enc)
+tcp.close(connection: conn)
```

Comment:
I like the general pattern here of passing raw data to transports. I think this will work well.

The pattern of opening a connection and then closing it is very imperical and not very functional friendly, we should probably rethink that pattern. That said `tcp` is not part of the current requirements so we can punt on that design if it proves to be very challenging.